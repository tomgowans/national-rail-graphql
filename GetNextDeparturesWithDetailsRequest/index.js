const fetchData = require("../fetchData");

/**
 *
 * @param {number || null} numRows (integer, between 0 and 150 exclusive): The number of services to return in the resulting station board.
 * @param {string} crs (string, 3 characters, alphabetic): The CRS code (see above) of the location for which the request is being made.
 * @param {string || null} filterCrs (string, 3 characters, alphabetic): The CRS code of either an origin or destination location to filter in. Optional.
 * @param {string || null} filterType (string, either "from" or "to"): The type of filter to apply. Filters services to include only those originating or terminating at the filterCrs location. Defaults to "to". Optional.
 * @param {number || null} timeOffset (integer, between -120 and 120 exclusive): An offset in minutes against the current time to provide the station board for. Defaults to 0. Optional.
 * @param {number || null} timeWindow (integer, between -120 and 120 exclusive): How far into the future in minutes, relative to timeOffset, to return services for. Defaults to 120. Optional.
 */
function GetNextDeparturesWithDetailsRequest(
  {
    numRows = null,
    crs,
    filterCrs = [],
    filterType = "to",
    timeOffset = 0,
    timeWindow = 120
  },
  tokenValue
) {
  return fetchData({
    body: `
      <ldb:GetNextDeparturesWithDetailsRequest>
      <ldb:crs>${crs}</ldb:crs>
      ${filterCrs.length &&
        `<ldb:filterList>
        ${filterCrs.map(filterCrsItem => `<ldb:crs>${filterCrsItem}</ldb:crs>`)}
      </ldb:filterList>`}
      <ldb:timeOffset>${timeOffset}</ldb:timeOffset>
      <ldb:timeWindow>${timeWindow}</ldb:timeWindow>
    </ldb:GetNextDeparturesWithDetailsRequest>

      `,
    tokenValue
  })
    .then(result => {
      return result.GetNextDeparturesWithDetailsResponse[0].DeparturesBoard[0];
      // console.dir(
      //   result.GetNextDeparturesWithDetailsResponse[0].DeparturesBoard[0]
      //     .departures[0].destination[0].service[0]
      // );
    })
    .catch(err => console.error(err));
}

module.exports = GetNextDeparturesWithDetailsRequest;
