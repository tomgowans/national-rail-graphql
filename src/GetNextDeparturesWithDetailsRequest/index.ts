import fetchData from "../fetchData";
import removeArrays from "../removeArrays";
import { Attributes, Result } from "../types";

/**
 *
 * @param {string} crs (string, 3 characters, alphabetic): The CRS code (see above) of the location for which the request is being made.
 * @param {string || null} filterCrs (string, 3 characters, alphabetic): The CRS code of either an origin or destination location to filter in. Optional.
 * @param {number || null} timeOffset (integer, between -120 and 120 exclusive): An offset in minutes against the current time to provide the station board for. Defaults to 0. Optional.
 * @param {number || null} timeWindow (integer, between -120 and 120 exclusive): How far into the future in minutes, relative to timeOffset, to return services for. Defaults to 120. Optional.
 */

function GetNextDeparturesWithDetailsRequest(
  { crs, filterCrs = [], timeOffset = 0, timeWindow = 120 }: Attributes,
  tokenValue: string
) {
  return new Promise((resolve, reject) => {
    fetchData({
      body: `
      <ldb:GetNextDeparturesWithDetailsRequest>
      <ldb:crs>${crs}</ldb:crs>
      ${filterCrs.length &&
        `<ldb:filterList>
        ${filterCrs.map(filterCrsItem => `<ldb:crs>${filterCrsItem}</ldb:crs>`)}
      </ldb:filterList>`}
      <ldb:timeOffset>${timeOffset}</ldb:timeOffset>
      <ldb:timeWindow>${timeWindow}</ldb:timeWindow>
    </ldb:GetNextDeparturesWithDetailsRequest>

      `,
      tokenValue
    })
      .then((result: Result) => {
        // console.dir(result);
        resolve(
          removeArrays(
            result.GetNextDeparturesWithDetailsResponse[0].DeparturesBoard[0]
          )
        );
      })
      .catch(err => reject(err));
  });
}

export default GetNextDeparturesWithDetailsRequest;
