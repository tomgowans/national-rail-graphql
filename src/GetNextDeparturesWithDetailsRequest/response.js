const response = {
  text: `<?xml version="1.0" encoding="utf-8"?>
    <soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
        <soap:Body>
            <GetNextDeparturesWithDetailsResponse xmlns="http://thalesgroup.com/RTTI/2017-10-01/ldb/">
                <DeparturesBoard xmlns:lt="http://thalesgroup.com/RTTI/2012-01-13/ldb/types" xmlns:lt6="http://thalesgroup.com/RTTI/2017-02-02/ldb/types" xmlns:lt7="http://thalesgroup.com/RTTI/2017-10-01/ldb/types" xmlns:lt4="http://thalesgroup.com/RTTI/2015-11-27/ldb/types" xmlns:lt5="http://thalesgroup.com/RTTI/2016-02-16/ldb/types" xmlns:lt2="http://thalesgroup.com/RTTI/2014-02-20/ldb/types" xmlns:lt3="http://thalesgroup.com/RTTI/2015-05-14/ldb/types">
                    <lt4:generatedAt>2019-12-17T17:36:59.1755832+00:00</lt4:generatedAt>
                    <lt4:locationName>East Croydon</lt4:locationName>
                    <lt4:crs>ECR</lt4:crs>
                    <lt4:nrccMessages>
                        <lt:message>Disruption between London and Stevenage. More details can be found in &lt;A href="http://nationalrail.co.uk/service_disruptions/238310.aspx"&gt;Latest Travel News.&lt;/A&gt;</lt:message>
                        <lt:message>Disruption through Redhill. More information can be found in &lt;A href="http://nationalrail.co.uk/service_disruptions/238336.aspx"&gt;Latest Travel News.&lt;/A&gt;</lt:message>
                    </lt4:nrccMessages>
                    <lt4:platformAvailable>true</lt4:platformAvailable>
                    <lt7:departures>
                        <lt7:destination crs="STP">
                            <lt7:service>
                                <lt4:sta>17:31</lt4:sta>
                                <lt4:eta>17:40</lt4:eta>
                                <lt4:std>17:31</lt4:std>
                                <lt4:etd>17:41</lt4:etd>
                                <lt4:platform>1</lt4:platform>
                                <lt4:operator>Thameslink</lt4:operator>
                                <lt4:operatorCode>TL</lt4:operatorCode>
                                <lt4:serviceType>train</lt4:serviceType>
                                <lt4:serviceID>4dU8fydGLLMrBhZUs9E2Uw==</lt4:serviceID>
                                <lt5:origin>
                                    <lt4:location>
                                        <lt4:locationName>Brighton</lt4:locationName>
                                        <lt4:crs>BTN</lt4:crs>
                                    </lt4:location>
                                </lt5:origin>
                                <lt5:destination>
                                    <lt4:location>
                                        <lt4:locationName>Cambridge</lt4:locationName>
                                        <lt4:crs>CBG</lt4:crs>
                                    </lt4:location>
                                </lt5:destination>
                                <lt7:subsequentCallingPoints>
                                    <lt7:callingPointList>
                                        <lt7:callingPoint>
                                            <lt7:locationName>London Bridge</lt7:locationName>
                                            <lt7:crs>LBG</lt7:crs>
                                            <lt7:st>17:45</lt7:st>
                                            <lt7:et>17:54</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>London Blackfriars</lt7:locationName>
                                            <lt7:crs>BFR</lt7:crs>
                                            <lt7:st>17:51</lt7:st>
                                            <lt7:et>18:00</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>City Thameslink</lt7:locationName>
                                            <lt7:crs>CTK</lt7:crs>
                                            <lt7:st>17:53</lt7:st>
                                            <lt7:et>18:02</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Farringdon Underground</lt7:locationName>
                                            <lt7:crs>ZFD</lt7:crs>
                                            <lt7:st>17:56</lt7:st>
                                            <lt7:et>18:04</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>London St Pancras (Intl)</lt7:locationName>
                                            <lt7:crs>STP</lt7:crs>
                                            <lt7:st>18:00</lt7:st>
                                            <lt7:et>18:08</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Finsbury Park</lt7:locationName>
                                            <lt7:crs>FPK</lt7:crs>
                                            <lt7:st>18:07</lt7:st>
                                            <lt7:et>18:15</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Stevenage</lt7:locationName>
                                            <lt7:crs>SVG</lt7:crs>
                                            <lt7:st>18:28</lt7:st>
                                            <lt7:et>18:35</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Hitchin</lt7:locationName>
                                            <lt7:crs>HIT</lt7:crs>
                                            <lt7:st>18:35</lt7:st>
                                            <lt7:et>18:40</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Letchworth Garden City</lt7:locationName>
                                            <lt7:crs>LET</lt7:crs>
                                            <lt7:st>18:41</lt7:st>
                                            <lt7:et>18:46</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Baldock</lt7:locationName>
                                            <lt7:crs>BDK</lt7:crs>
                                            <lt7:st>18:45</lt7:st>
                                            <lt7:et>18:49</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Ashwell &amp; Morden</lt7:locationName>
                                            <lt7:crs>AWM</lt7:crs>
                                            <lt7:st>18:50</lt7:st>
                                            <lt7:et>18:54</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Royston</lt7:locationName>
                                            <lt7:crs>RYS</lt7:crs>
                                            <lt7:st>18:54</lt7:st>
                                            <lt7:et>18:58</lt7:et>
                                        </lt7:callingPoint>
                                        <lt7:callingPoint>
                                            <lt7:locationName>Cambridge</lt7:locationName>
                                            <lt7:crs>CBG</lt7:crs>
                                            <lt7:st>19:10</lt7:st>
                                            <lt7:et>19:13</lt7:et>
                                        </lt7:callingPoint>
                                    </lt7:callingPointList>
                                </lt7:subsequentCallingPoints>
                            </lt7:service>
                        </lt7:destination>
                    </lt7:departures>
                </DeparturesBoard>
            </GetNextDeparturesWithDetailsResponse>
        </soap:Body>
    </soap:Envelope>`,
};

module.exports = response;
