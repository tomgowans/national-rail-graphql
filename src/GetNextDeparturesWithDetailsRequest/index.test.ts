import GetNextDeparturesWithDetailsRequest from "./index";

describe("GetNextDeparturesWithDetailsRequest", () => {
  it("sends a request with minimal parameters correctly", async () => {
    expect.assertions(1);

    const result = await GetNextDeparturesWithDetailsRequest(
      {
        crs: "ECR"
      },
      "TOKEN_VALUE"
    );
    expect(result).toMatchSnapshot();
  });

  it("sends a request with minimal parameters and fail", async () => {
    expect.assertions(1);

    await expect(
      GetNextDeparturesWithDetailsRequest(
        {
          crs: "ECR"
        },
        ""
      )
    ).rejects.toThrow("Error with credentials");
  });

  it("sends a request with more parameters correctly", () => {
    expect.assertions(1);

    return expect(
      GetNextDeparturesWithDetailsRequest(
        {
          numRows: 12,
          crs: "ECR",
          filterCrs: ["STP"]
        },
        "TOKEN_VALUE"
      )
    ).resolves.toMatchSnapshot();
  });
});
