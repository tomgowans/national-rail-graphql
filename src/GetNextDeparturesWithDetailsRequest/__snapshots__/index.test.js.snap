// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetNextDeparturesWithDetailsRequest sends a request with minimal parameters correctly 1`] = `
Object {
  "$": Object {
    "xmlns:lt": "http://thalesgroup.com/RTTI/2012-01-13/ldb/types",
    "xmlns:lt2": "http://thalesgroup.com/RTTI/2014-02-20/ldb/types",
    "xmlns:lt3": "http://thalesgroup.com/RTTI/2015-05-14/ldb/types",
    "xmlns:lt4": "http://thalesgroup.com/RTTI/2015-11-27/ldb/types",
    "xmlns:lt5": "http://thalesgroup.com/RTTI/2016-02-16/ldb/types",
    "xmlns:lt6": "http://thalesgroup.com/RTTI/2017-02-02/ldb/types",
    "xmlns:lt7": "http://thalesgroup.com/RTTI/2017-10-01/ldb/types",
  },
  "crs": "ECR",
  "departures": Object {
    "destination": Object {
      "$": Object {
        "crs": "STP",
      },
      "service": Array [
        Object {
          "destination": Object {
            "location": Array [
              Object {
                "crs": "CBG",
                "locationName": "Cambridge",
              },
            ],
          },
          "eta": "17:40",
          "etd": "17:41",
          "operator": "Thameslink",
          "operatorCode": "TL",
          "origin": Object {
            "location": Array [
              Object {
                "crs": "BTN",
                "locationName": "Brighton",
              },
            ],
          },
          "platform": 1,
          "serviceID": "4dU8fydGLLMrBhZUs9E2Uw==",
          "serviceType": "train",
          "sta": "17:31",
          "std": "17:31",
          "subsequentCallingPoints": Object {
            "callingPointList": Object {
              "callingPoint": Array [
                Object {
                  "crs": "LBG",
                  "et": "17:54",
                  "locationName": "London Bridge",
                  "st": "17:45",
                },
                Object {
                  "crs": "BFR",
                  "et": "18:00",
                  "locationName": "London Blackfriars",
                  "st": "17:51",
                },
                Object {
                  "crs": "CTK",
                  "et": "18:02",
                  "locationName": "City Thameslink",
                  "st": "17:53",
                },
                Object {
                  "crs": "ZFD",
                  "et": "18:04",
                  "locationName": "Farringdon Underground",
                  "st": "17:56",
                },
                Object {
                  "crs": "STP",
                  "et": "18:08",
                  "locationName": "London St Pancras (Intl)",
                  "st": "18:00",
                },
                Object {
                  "crs": "FPK",
                  "et": "18:15",
                  "locationName": "Finsbury Park",
                  "st": "18:07",
                },
                Object {
                  "crs": "SVG",
                  "et": "18:35",
                  "locationName": "Stevenage",
                  "st": "18:28",
                },
                Object {
                  "crs": "HIT",
                  "et": "18:40",
                  "locationName": "Hitchin",
                  "st": "18:35",
                },
                Object {
                  "crs": "LET",
                  "et": "18:46",
                  "locationName": "Letchworth Garden City",
                  "st": "18:41",
                },
                Object {
                  "crs": "BDK",
                  "et": "18:49",
                  "locationName": "Baldock",
                  "st": "18:45",
                },
                Object {
                  "crs": "AWM",
                  "et": "18:54",
                  "locationName": "Ashwell & Morden",
                  "st": "18:50",
                },
                Object {
                  "crs": "RYS",
                  "et": "18:58",
                  "locationName": "Royston",
                  "st": "18:54",
                },
                Object {
                  "crs": "CBG",
                  "et": "19:13",
                  "locationName": "Cambridge",
                  "st": "19:10",
                },
              ],
            },
          },
        },
      ],
    },
  },
  "generatedAt": "2019-12-17T17:36:59.1755832+00:00",
  "locationName": "East Croydon",
  "nrccMessages": Object {
    "message": Array [
      "Disruption between London and Stevenage. More details can be found in <A href=\\"http://nationalrail.co.uk/service_disruptions/238310.aspx\\">Latest Travel News.</A>",
      "Disruption through Redhill. More information can be found in <A href=\\"http://nationalrail.co.uk/service_disruptions/238336.aspx\\">Latest Travel News.</A>",
    ],
  },
  "platformAvailable": true,
}
`;

exports[`GetNextDeparturesWithDetailsRequest sends a request with more parameters correctly 1`] = `
Object {
  "$": Object {
    "xmlns:lt": "http://thalesgroup.com/RTTI/2012-01-13/ldb/types",
    "xmlns:lt2": "http://thalesgroup.com/RTTI/2014-02-20/ldb/types",
    "xmlns:lt3": "http://thalesgroup.com/RTTI/2015-05-14/ldb/types",
    "xmlns:lt4": "http://thalesgroup.com/RTTI/2015-11-27/ldb/types",
    "xmlns:lt5": "http://thalesgroup.com/RTTI/2016-02-16/ldb/types",
    "xmlns:lt6": "http://thalesgroup.com/RTTI/2017-02-02/ldb/types",
    "xmlns:lt7": "http://thalesgroup.com/RTTI/2017-10-01/ldb/types",
  },
  "crs": "ECR",
  "departures": Object {
    "destination": Object {
      "$": Object {
        "crs": "STP",
      },
      "service": Array [
        Object {
          "destination": Object {
            "location": Array [
              Object {
                "crs": "CBG",
                "locationName": "Cambridge",
              },
            ],
          },
          "eta": "17:40",
          "etd": "17:41",
          "operator": "Thameslink",
          "operatorCode": "TL",
          "origin": Object {
            "location": Array [
              Object {
                "crs": "BTN",
                "locationName": "Brighton",
              },
            ],
          },
          "platform": 1,
          "serviceID": "4dU8fydGLLMrBhZUs9E2Uw==",
          "serviceType": "train",
          "sta": "17:31",
          "std": "17:31",
          "subsequentCallingPoints": Object {
            "callingPointList": Object {
              "callingPoint": Array [
                Object {
                  "crs": "LBG",
                  "et": "17:54",
                  "locationName": "London Bridge",
                  "st": "17:45",
                },
                Object {
                  "crs": "BFR",
                  "et": "18:00",
                  "locationName": "London Blackfriars",
                  "st": "17:51",
                },
                Object {
                  "crs": "CTK",
                  "et": "18:02",
                  "locationName": "City Thameslink",
                  "st": "17:53",
                },
                Object {
                  "crs": "ZFD",
                  "et": "18:04",
                  "locationName": "Farringdon Underground",
                  "st": "17:56",
                },
                Object {
                  "crs": "STP",
                  "et": "18:08",
                  "locationName": "London St Pancras (Intl)",
                  "st": "18:00",
                },
                Object {
                  "crs": "FPK",
                  "et": "18:15",
                  "locationName": "Finsbury Park",
                  "st": "18:07",
                },
                Object {
                  "crs": "SVG",
                  "et": "18:35",
                  "locationName": "Stevenage",
                  "st": "18:28",
                },
                Object {
                  "crs": "HIT",
                  "et": "18:40",
                  "locationName": "Hitchin",
                  "st": "18:35",
                },
                Object {
                  "crs": "LET",
                  "et": "18:46",
                  "locationName": "Letchworth Garden City",
                  "st": "18:41",
                },
                Object {
                  "crs": "BDK",
                  "et": "18:49",
                  "locationName": "Baldock",
                  "st": "18:45",
                },
                Object {
                  "crs": "AWM",
                  "et": "18:54",
                  "locationName": "Ashwell & Morden",
                  "st": "18:50",
                },
                Object {
                  "crs": "RYS",
                  "et": "18:58",
                  "locationName": "Royston",
                  "st": "18:54",
                },
                Object {
                  "crs": "CBG",
                  "et": "19:13",
                  "locationName": "Cambridge",
                  "st": "19:10",
                },
              ],
            },
          },
        },
      ],
    },
  },
  "generatedAt": "2019-12-17T17:36:59.1755832+00:00",
  "locationName": "East Croydon",
  "nrccMessages": Object {
    "message": Array [
      "Disruption between London and Stevenage. More details can be found in <A href=\\"http://nationalrail.co.uk/service_disruptions/238310.aspx\\">Latest Travel News.</A>",
      "Disruption through Redhill. More information can be found in <A href=\\"http://nationalrail.co.uk/service_disruptions/238336.aspx\\">Latest Travel News.</A>",
    ],
  },
  "platformAvailable": true,
}
`;
